### PR TITLE
add support for messages for meter

### LINKED ISSUE #7326: Meter - screenreader accessibility
### Expected Behavior
Meter information should be read out for a screenreader user.

### Actual Behavior
Currently, Voiceover doesn't have access to read chart text information; just read out as "image" with no content information.

### URL, screen shot, or Codepen exhibiting the issue
This issue can be seen in any of the Meter storybook stories
https://storybook.grommet.io/?path=/story/visualizations-meter-bar--bar

### Steps to Reproduce

1. 2. 3.

### Your Environment

- Grommet version:
- Browser Name and version:
- Operating System and version (desktop or mobile):

### PR DESCRIPTION
**Pull request by britt6612**
#### What does this PR do?
adds `messages` to be passed through the `Grommet` tag for Meter component
#### Where should the reviewer start?
meter.js
#### What testing has been done on this PR?
storybook
#### How should this be manually tested?
storybook
#### Do Jest tests follow these best practices?
TODO:

- [ ] `screen` is used for querying.
- [ ] The correct query is used. (Refer to [this list of queries](https://testing-library.com/docs/queries/about/#priority))
- [ ] `asFragment()` is used for snapshot testing.

#### Any background context you want to provide?
previously screen reader just read out that our meter was an image. We need to provide better context
#### What are the relevant issues?
closes #7326 
#### Do the grommet docs need to be updated?
yes
#### Should this PR be mentioned in the release notes?
yes
#### Is this change backwards compatible or is it a breaking change?
backwards compatible

### CODE REVIEW

File: src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap [hunk 1]
```diff
@@ -24,6 +24,7 @@ exports[`Meter background 1`] = `
   class="c0"
 >
   <svg
+    aria-label="bar meter with undefined% filled"
```
Comment (MikeKingdom) [inline src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap]: What's happening with all the `undefined` substitution values?

File: src/js/components/Meter/Meter.js [hunk 2]
```diff
@@ -46,12 +50,41 @@ const Meter = forwardRef(
...
+      ariaLabel ||
+      format({
+        id: meterType,
+        messages: messages?.meter,
+        values: {
+          percentage: value,
+          type,
+        },
+      });
+
```
Comment (MikeKingdom) [inline src/js/components/Meter/Meter.js]:
```suggestion
    const meterAriaLabel =
      ariaLabel ||
      format({
        id: `meter.${type ?? 'default'}`,
        messages: messages?.meter,
        values: {
          percentage: value ?? 0,
          type,
        },
      });

```

**Review by MikeKingdom — COMMENTED** [review]
Needs the messages property in Meter/index.d.ts and propTypes.js

File: src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap (hunk 1 above)
Comment (britt6612) [reply src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap]: its when there is no `value` fixing now

Comment (britt6612) [reply src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap]: in one test we have <Meter /> without any `value` passed in

Comment (britt6612) [reply src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap]: There shouldnt be anymore undefined

File: src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap [hunk 3]
```diff
@@ -249,6 +256,7 @@ exports[`Meter boundary values 1`] = `
   class="c0"
 >
   <svg
+    aria-label="circle meter with value 2.26% out of 10.299999999999999"
```
Comment (britt6612) [inline src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap:259]: @jcfilben still not sure about the wording.. what you think

Comment (britt6612) [reply src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap:259]: okay I wonder if we take out the `%` and just have the user pass in whatever units they want

Comment (jcfilben) [reply src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap]: Yeah I think we should take out the `%`

Comment (britt6612) [reply src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap]: agree took out in last commit

File: src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap [hunk 4]
```diff
@@ -249,6 +256,7 @@ exports[`Meter boundary values 1`] = `
   class="c0"
 >
   <svg
+    aria-label="circle meter with value 2.26 out of 10.299999999999999"
```
[the remark this answers is not in the document]
Comment (britt6612) [reply src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap:259]: I can add the 8.04 in the message in the plural one..

so the max is coming from here: 
```
const deriveMax = (values) => {
  let max = 100;
  if (values && values.length > 1) {
    max = 0;
    values.forEach((v) => {
      max += v.value;
    });
  }
  return max;
};
```

`    const memoizedMax = useMemo(() => deriveMax(values), [values]);`

the `memoizedMax` is what Im passing through to the message.
if a user was passing `max` this was being put through the `...rest` so we didnt have access to this value pulling out the `max` so we can use that if it is being passed through

File: src/js/languages/default.json [hunk 5]
```diff
@@ -104,6 +104,10 @@
     "openMenu": "Open Menu",
     "closeMenu": "Close Menu"
   },
+  "meter": {
+    "singular": "{type} meter with value {meterValue} filled out of {max}",
+    "plural": "{type} meter with values {meterValue} filled out of {max}"
```
Comment (jcfilben) [inline src/js/languages/default.json:109]:
```suggestion
    "singular": "{type} meter with value {meterValue} out of {max}",
    "plural": "{type} meter with values {meterValue} out of {max}"
```

File: src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap [hunk 6]
```diff
@@ -306,6 +364,7 @@ exports[`Meter default 1`] = `
   class="c0"
 >
   <svg
+    aria-label="bar meter with values  filled out of 100"
```
Comment (jcfilben) [inline src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap:367]: Can we have a message that handles when there is no value?

**Comment by jcfilben** [comment]
Looks like the max bundlesize value needs bumped up a little

File: src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap (hunk 6 above)
Comment (britt6612) [reply src/js/components/Meter/__tests__/__snapshots__/Meter-test.tsx.snap]: done

**Review by jcfilben — APPROVED** [review]
Looks good!

File: src/js/components/Meter/Meter.js [hunk 7]
```diff
@@ -19,9 +20,12 @@ const deriveMax = (values) => {
 const Meter = forwardRef(
   (
     {
+      'aria-label': ariaLabel,
       background = { color: 'light-2', opacity: 'medium' },
       color,
       direction = 'horizontal',
+      max,
```
Comment (halocline) [inline src/js/components/Meter/Meter.js:27]: Was `max` documented, but not an actual prop?

Comment (halocline) [inline src/js/components/Meter/Meter.js:27]: `max` documentation states "The maximum value for the Meter." Does this imply that if I specify `max={10}`, then set `value={20}`, the 20 won't be respected? If not, should the definition of max be clarified?

Comment (halocline) [reply src/js/components/Meter/Meter.js:27]: Are there other props documented but not represented as props here?

File: src/js/components/Meter/Meter.js [hunk 8]
```diff
@@ -47,12 +52,28 @@ const Meter = forwardRef(
...
+      });
+
     let content;
     if (type === 'bar') {
       content = (
         <Bar
           ref={ref}
-          max={memoizedMax}
+          aria-label={meterAriaLabel}
+          max={max || memoizedMax}
```
Comment (MikeKingdom) [inline src/js/components/Meter/Meter.js:76]: This isn't quite the same but probably an improvement.
Previously if the max prop was in `rest` I think it would override this property. Now you will get `memoizedMax` if max is specified but falsey (e.g. 0, undefined, null). Maybe that's better? That does assume we can't have negative values (e.g. say a range of values from -10 to 0)?

I think it's little fragile that the rules for what max is are now split into 4 places:

1. deriveMax
2. the message max param
3. Bar max property (does logic)
4. Cirlce max property (does logic)

This is alot of risk for disagreement. Right now if there are multiple `values` but they are 0 you will get a `memoizedMax` of 0. The message will say 100 but the Bar and Circle will be passed `max=0`

`memoizedMax` will be 100 or if there are at least 2 `values` it adds them up. That also means if `values` only contains 1 thing it will return 100 or if there's a `value` it will return 100.

To avoid this we might want to refactor to normalize max in one spot and pass that one value to the message, Bar and Circle. Maybe something like:

Change the component prop to be `max: maxProp` and then
```javascript
const max = useMemo(() => {
  let maxValue = 100;
  if (values?.length > 1) {
    maxValue = values.reduce((total, currentValue) => total + currentValue, 0);
  }
  return maxValue || maxProp || 100;
}, [maxProp, values]);  
```
 And then just use `max` for the 3 places it's needed (message param, Bar prop and Circle prop).

File: src/js/components/Meter/__tests__/Meter-test.tsx [hunk 9]
```diff
@@ -248,4 +248,22 @@ describe('Meter', () => {
...
   });
+
+  test('custom message', () => {
+    const value = 30;
+    const { container } = render(
+      <Grommet
+        messages={{
+          messages: {
+            meter: {
+              singular: `bar value is ${value}% out of 100.`,
```
[the remark this answers is not in the document]
Comment (britt6612) [reply src/js/components/Meter/__tests__/Meter-test.tsx:259]: adding.

File: src/js/components/Meter/__tests__/Meter-test.tsx [hunk 10]
```diff
@@ -248,4 +248,22 @@ describe('Meter', () => {
...
+            },
+          },
+        }}
+      >
+        <Meter value={value} />
```
[the remark this answers is not in the document]
Comment (britt6612) [reply src/js/components/Meter/__tests__/Meter-test.tsx:264]: yes correct a user can add on the Meter directly adding

File: src/js/components/Meter/Meter.js (hunk 7 above)
Comment (britt6612) [reply src/js/components/Meter/Meter.js:27]: yes there are other props documented but not represented here I had pulled out the `max` so I could use that in the message since that was just being passed in the `...rest`

Comment (britt6612) [reply src/js/components/Meter/Meter.js:27]: correct we can clarify our documentation. 

so if we have the following: 
```
      <Meter type="bar" max={50} value={60} />
```

The whole meter will be filled. 

Screenshot 2024-11-25 at 7 18 01 PM

File: src/js/components/Meter/Meter.js (hunk 8 above)
Comment (britt6612) [reply src/js/components/Meter/Meter.js]: @MikeKingdom 

yes agree there is alot being done here 

```
deriveMax
the message max param
Bar max property (does logic)
Cirlce max property (does logic)
```

I added the suggestion you provided with a small change. 

1. I needed 
` maxValue = values.reduce((total, currentValue) => total + currentValue, 0);`
currentValue.value

2. I put the `maxProp` to be the first in our || statement so its the following 
`  return maxProp || maxValue || 100;`

for example if I have 
```
      <Meter type="bar" max={40} value={30} />
```
with having maxValue || maxProp || 100 
it was taking the maxValue of 100 and not respecting what the user passed in for `maxProp`

File: src/js/languages/default.json [hunk 11]
```diff
@@ -104,6 +104,10 @@
     "openMenu": "Open Menu",
     "closeMenu": "Close Menu"
   },
+  "meter": {
+    "singular": "{type} meter with value {meterValue} out of {max}",
+    "plural": "{type} meter with values {meterValue} out of {max}"
```
Comment (MikeKingdom) [inline src/js/languages/default.json:109]: One thing I just realized is the `type` isn't currently localizable. We should have keys in here for type. Although a bit wordy, I think a more flexible thing for localizers would be:
```suggestion
    "bar": {
      "singular": "Bar meter with value {meterValue} out of {max}",
      "plural": "Bar meter with values {meterValue} out of {max}"
    },
    "circle": {
      "singular": "Circle meter with value {meterValue} out of {max}",
      "plural": "Circle meter with values {meterValue} out of {max}"
    }
```

Comment (britt6612) [reply src/js/languages/default.json:109]: aw okay good point I can restructure

Comment (britt6612) [reply src/js/languages/default.json]: @MikeKingdom updated structure

**Review by MikeKingdom — APPROVED** [review]
Thanks!

**Review by jcfilben — APPROVED** [review]
Looks good!
